src/providers/wms/qgswmsdataitems.cpp

```diff
@@ -233,31 +237,87 @@ bool QgsWMSConnectionItem::equal( const QgsDataItem *other )
 
 // ---------------------------------------------------------------------------
 
-QgsWMSLayerItem::QgsWMSLayerItem( QgsDataItem *parent, QString name, QString path, const QgsWmsCapabilitiesProperty &capabilitiesProperty, const QgsDataSourceUri &dataSourceUri, const QgsWmsLayerProperty &layerProperty )
-  : QgsLayerItem( parent, name, path, QString(), QgsLayerItem::Raster, QStringLiteral( "wms" ) )
+QgsWMSLayerCollectionItem::QgsWMSLayerCollectionItem( QgsDataItem *parent, QString name, QString path, const QgsWmsCapabilitiesProperty &capabilitiesProperty, const QgsDataSourceUri &dataSourceUri, const QgsWmsLayerProperty &layerProperty )
+  : QgsDataCollectionItem( parent, name, path )
   , mCapabilitiesProperty( capabilitiesProperty )
   , mDataSourceUri( dataSourceUri )
   , mLayerProperty( layerProperty )
 {
-  mSupportedCRS = mLayerProperty.crs;
-  mSupportFormats = mCapabilitiesProperty.capability.request.getMap.format;
-  QgsDebugMsgLevel( "uri = " + mDataSourceUri.encodedUri(), 2 );
-
-  mUri = createUri();
+  mIconName = QStringLiteral( "mIconWms.svg" );
 
   // Populate everything, it costs nothing, all info about layers is collected
   for ( const QgsWmsLayerProperty &layerProperty : qgis::as_const( mLayerProperty.layer ) )
   {
     // Attention, the name may be empty
     QgsDebugMsgLevel( QString::number( layerProperty.orderId ) + ' ' + layerProperty.name + ' ' + layerProperty.title, 2 );
     QString pathName = layerProperty.name.isEmpty() ? QString::number( layerProperty.orderId ) : layerProperty.name;
+
     QgsWMSLayerItem *layer = new QgsWMSLayerItem( this, layerProperty.title, mPath + '/' + pathName, mCapabilitiesProperty, dataSourceUri, layerProperty );
-    //mChildren.append( layer );
+
     addChildItem( layer );
   }
 
-  mIconName = QStringLiteral( "mIconWms.svg" );
+  setState( Populated );
+}
+
+bool QgsWMSLayerCollectionItem::equal( const QgsDataItem *other )
+{
+  if ( type() != other->type() )
+  {
+    return false;
+  }
+  const QgsWMSLayerCollectionItem *otherCollectionItem = dynamic_cast<const QgsWMSLayerCollectionItem *>( other );
```

Comment:
```suggestion
  const QgsWMSLayerCollectionItem *otherCollectionItem = qobject_cast<const QgsWMSLayerCollectionItem *>( other );
```

Reply:
Right, does this also apply here?https://github.com/qgis/QGIS/blob/3bdb75bda4e61dd397b0cc553baa67035fd1a25a/src/providers/wms/qgswmsdataitems.cpp#L376

Reply:
Yep. qobject_cast is optimised and should be used whenever casting qobject subclasses. See https://stackoverflow.com/a/43994766/1861260

Reply:
Ooh I see